Fix Sonos speaker async_offline assertion failure

ptarjan opened the pull request:
Handle the case where async_offline() is called before async_subscribe() has been invoked, which would leave _subscription_lock as None. This can happen during shutdown when a speaker was discovered but no activity was detected yet. Instead of asserting, create the lock if it doesn't exist, matching the existing pattern in async_subscribe().

Failed in: https://github.com/home-assistant/core/actions/runs/20151776328/job/57846433236

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Root cause: In SonosSpeaker:

* _subscription_lock is initialized as None in __init__ (line 133)
* The lock is only created when async_subscribe() is called (lines 374-375)
* async_offline() asserts the lock is not None
* During shutdown, async_offline() is called on all discovered speakers, including those that never had async_subscribe() called (e.g., speakers discovered but with no activity detected yet)
* Fix (homeassistant/components/sonos/speaker.py:691-695): Instead of asserting, create the lock if it doesn't exist - matching the existing pattern in async_subscribe():




## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [ ] I understand the code I am submitting and can explain how it works.
- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.
- [ ] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

ptarjan committed:
Fix Sonos speaker async_offline assertion failure

Handle the case where async_offline() is called before async_subscribe()
has been invoked, which would leave _subscription_lock as None. This can
happen during shutdown when a speaker was discovered but no activity was
detected yet. Instead of asserting, create the lock if it doesn't exist,
matching the existing pattern in async_subscribe().

ptarjan committed:
Add test for async_offline without subscription lock

Tests the fix for the AssertionError when async_offline() is called
on a speaker where _subscription_lock is None (e.g., when a speaker
is discovered but no activity was detected before shutdown).

PeteRager commented on tests/components/sonos/test_speaker.py:
```diff
@@ -189,3 +189,34 @@ async def test_zgs_avtransport_group_speakers(
+    speaker = list(config_entry.runtime_data.discovered.values())[0]
+
+    # Ensure _subscription_lock is None (no activity triggered yet)
+    speaker._subscription_lock = None
```
We should not be directly manipulating integration internals from tests

PeteRager reviewed (changes requested):
Tests should not access internals

PeteRager commented on tests/components/sonos/test_speaker.py:
```diff
@@ -189,3 +189,30 @@ async def test_zgs_avtransport_group_speakers(
+    with patch(
+        "homeassistant.components.sonos.speaker.SonosSpeaker.async_subscribe",
+        new_callable=AsyncMock,
+    ):
```
We are not allowed to patch the integration, the only point that can be patched are in external libraries (for example SoCo)

ptarjan committed:
Add wait_background_tasks=True to test for proper async handling

Ensure all background tasks complete before checking results in the
async_offline_without_subscription_lock test.

ptarjan committed:
Patch async_setup instead of async_subscribe for better test isolation

Patching async_setup ensures the subscription lock is never created,
since async_subscribe (which creates the lock) is called from async_setup.

ptarjan committed:
Use SoCo mock to test async_offline without subscription lock

Instead of patching the integration, configure the SoCo mock's play_mode
property to raise SoCoException. This causes setup() to fail early
(before scheduling async_setup), leaving the speaker in discovered
without _subscription_lock being created.

ptarjan committed:
Format test file with ruff

PeteRager reviewed (approved):
Thank you!
